src/cheerio_crawler.js

```diff
@@ -302,29 +313,94 @@ class CheerioCrawler {
     }
 
     /**
-     * Default request function to be used.
+     * Default request function to be used. It performs optimizations
+     * on the request such as only downloading the request body if the
+     * received content type
      * @ignore
      */
     async _defaultRequestFunction({ request }) {
-        return rp(this._getRequestOptions(request));
+        return new Promise((resolve, reject) => {
+            // Using the streaming API of Request to be able to
+            // handle the response based on headers receieved.
+            this.rqst(this._getRequestOptions(request))
+                .on('error', err => reject(err))
+                .on('response', async (res) => {
```

Comment:
Anything that throws here below could result in unfinished promise that stucks forever and will be hard to debug. I think that we should make async function

```javascript
async _ensureResponseIsParseable(response) {
   ...
}
```

and then use it for example this way:

```javascript
.on('response', (res) => {
    this. _ensureResponseIsParseable(response, request)
          .then(resolve, reject);
}
```

Reply:
I don't think it will be stuck forever, because it should produce an uncaught exception that will throw in the callback scope and thus kill the process. Which was my intention, since I don't think the function should throw at any point, unless there's an error in my code.